File: BrainBrawlBackendPython/brain_brawl/remediation.py
from __future__ import annotations

from .types import DifficultyContext


class RemediationEngine:
    def next_difficulty(self, context: DifficultyContext) -> int:
        """
        Dynamic remediation:
        - Decrease difficulty if failures are high.
        - Increase difficulty when performance is strong and fast.
        - Clamp difficulty to 1..10.
        """
        next_value = context.current_difficulty

        if context.recent_failure_rate >= 0.45:
            next_value -= 1
        elif context.recent_failure_rate <= 0.15 and context.average_response_ms < 3500:
            next_value += 1

        if context.average_response_ms > 10_000:
            next_value -= 1

        return min(10, max(1, next_value))

File: BrainBrawlBackend/src/remediation.ts
import { DifficultyContext } from "./types";

export class RemediationEngine {
  /**
   * Dynamic remediation:
   * - Lower difficulty when failure rate climbs.
   * - Raise difficulty for high success + fast response.
   * - Always clamp to 1..10.
   */
  nextDifficulty(context: DifficultyContext): number {
    let next = context.currentDifficulty;

    if (context.recentFailureRate >= 0.45) {
      next -= 1;
    } else if (context.recentFailureRate <= 0.15 && context.averageResponseMs < 3500) {
      next += 1;
    }

    if (context.averageResponseMs > 10_000) {
      next -= 1;
    }

    return Math.min(10, Math.max(1, next));
  }
}
